### ISSUE TITLE
Query keys within an arbitrary shape?

### ISSUE THREAD

**Issue by janakagamini** [firebase/geofire-java#17]
This might be a stretch, but in addition to querying for keys within a radius, will it be possible (in the future perhaps) to query for keys within an arbitrary shape (possibly defined as a list of `GeoLocations`)?

E.g. usage: Check if a point is within a certain city (given that I know a list of GeoLocations that outlines the city).

**Comment by jwngr** [comment]
Thanks for the feature request! It is pretty much equivalent to https://github.com/firebase/geofire-js/issues/46. Unfortunately, I don't see this happening any time soon unless someone from the community wants to take a crack at it and send us a PR.

**Comment by janakagamini** [comment]
Yes, exactly that. I'll give it a try. From a cursory glance at the code it seems I should probably be looking at `private boolean locationIsInQuery(GeoLocation location)` @ [GeoQuery.java](https://github.com/firebase/geofire-java/blob/3b851b49883060bcafe833e76e2ab8a316a71310/src/main/java/com/firebase/geofire/GeoQuery.java#L110-L112)

Shout our if I'm way off. Thanks :)

**Comment by jwngr** [comment]
Yup, that is the right place. Thanks for taking a crack at this!

A couple notes:
* You will also need a new constructor since [the current one just takes a center and radius](https://github.com/firebase/geofire-java/blob/3b851b49883060bcafe833e76e2ab8a316a71310/src/main/java/com/firebase/geofire/GeoQuery.java#L103). You will need a constructor which takes in the coordinates for the arbitrary shape of our GeoQuery.
* You will need to update `locationIsInQuery()` to do something different depending on if it's a circular GeoQuery or a polygonal GeoQuery. Defining Firebase queries for a polygonal GeoQuery will be tough (which is why we haven't tackled it yet) and may end up be wildly inefficient, but hopefully you can figure out the math behind it!

**Comment by janakagamini** [comment]
Thanks for the tips jwngr, I will certainly try, but I can't guarantee anything.

**Comment by vanniktech** [comment]
The objective C implementaton for geofire has support for bounding boxes AFAIK.

**Comment by samtstern** [comment]
@vanniktech that's good to know.  We can use that code as a reference.

**Comment by samtstern** [comment]
I think it would be reasonable to try and support bounding boxes, but I don't think we should get into the math of doing arbitrary polygonal geo computations in this library.  It's probably much better for everyone to just use a bounding box and then use some client-side logic to narrow that down to the polygon.

**Comment by vanniktech** [comment]
Totally agree, bounding box and circular shape should be good enough.
